### PR TITLE
Features/issue10

### PR DESCRIPTION
**Pull request by wildarg**
EpisodeDownloader added

### CODE REVIEW

File: app/src/main/java/com/stepango/archetype/db/Contract.kt [hunk 1]
```diff
@@ -44,6 +44,16 @@ interface KeyValueRepo<Key : Any, Value : Any> {
     fun save(key: Key, value: Value): Single<Value>
 
     /**
+     * Give [value] by [key] on given [io.reactivex.Scheduler]
+     */
+    fun get(key: Key): Single<Value>
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/db/Contract.kt]: use `observe` and `observeAll` instead

File: app/src/main/java/com/stepango/archetype/player/di/Injector.kt [hunk 2]
```diff
@@ -6,22 +6,15 @@ import com.ninetyseconds.auckland.core.log.SimpleLogger
...
-import com.stepango.archetype.action.ApiAction
-import com.stepango.archetype.action.ApiActionHandler
-import com.stepango.archetype.action.ApiActionProducer
-import com.stepango.archetype.action.ContextAction
-import com.stepango.archetype.action.ContextActionHandler
-import com.stepango.archetype.action.ContextActionProducer
-import com.stepango.archetype.action.IdleAction
-import com.stepango.archetype.action.IntentMaker
-import com.stepango.archetype.action.IntentMakerImpl
+import com.stepango.archetype.action.*
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/di/Injector.kt]: remove imports folding

File: app/src/main/java/com/stepango/archetype/player/data/wrappers/Wrappers.kt [hunk 3]
```diff
@@ -13,14 +15,19 @@ data class EpisodeListItemWrapper(private val model: EpisodesModel) : StableId,
     override val stableId: Long = model.id
     val name: String = model.name
     val imageUrl: String = model.iconUrl
+    val state: EpisodeDownloadState = model.state
     override fun args(): Args = argsOf { episodeId { stableId } }
+    fun downloadButtonVisibility() = if (model.file == null) View.VISIBLE else View.GONE
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/data/wrappers/Wrappers.kt]: for visibility, there is special `visible` boolean binding

File: app/src/main/java/com/stepango/archetype/util/UriUtils.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import android.net.Uri
+
+/**
+ * Wild, 03.07.2017.
+ */
+
+fun getFileName(url: String): String {
+    val uri = Uri.parse(url)
+    var name = ""
+    for (path in uri.pathSegments)
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/util/UriUtils.kt]: use joinToString method

**Comment by stepango** [comment]
Please resolve codebeat issues.

File: app/src/main/java/com/stepango/archetype/player/loader/DownloadEpisodeAction.kt [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    val episodesRepo by lazyInject { episodesRepo() }
+
+    override fun invoke(context: Context, args: Args): Completable
+            = episodesRepo.getAll()
+            .subscribeOn(io())
+            .flatMapCompletable {
+                val file = File(context.filesDir, getFileName(it.audioUrl))
+                if (file.exists()) {
+                    logger.d("file exists for ${it.name}")
+                    it.file = file.absolutePath
```
Comment (nekdenis) [inline app/src/main/java/com/stepango/archetype/player/loader/DownloadEpisodeAction.kt]: episode model should be immutable.
make EpisodeModel be data class and then you can call it.copy{file = file.absolutePath}

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        return START_STICKY
+    }
+}
+
+class EpisodeLoader(val service: Service) : ProgressUpdateListener {
+
+    val repo by lazyInject { episodesRepo() }
+    val toaster by lazyInject { toaster() }
+
+    val queueSubject = PublishSubject.create<Long>()!!
```
Comment (nekdenis) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: avoid using !!. Just provide non nullable type (val queueSubject:PublishSubject<Long>)

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    val notificationHelper = NotificationHelper(service, hashCode())
+    val loader = ProgressOkLoader(this)
+    val waitStack = ArrayList<Long>()
+    val dispatcher = queueSubject
+            .timeout(10, TimeUnit.MINUTES)
+            .observeOn(Schedulers.io())
+            .subscribe(
+                    { load(it) },
+                    { stopLoading() }
+            )!!
```
Comment (nekdenis) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: avoid using !!

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+                .url(episode.audioUrl)
+                .build()
+
+        val response: Response = client.newCall(request).execute()
+        if (!response.isSuccessful) throw IOException("Unexpected code $response")
+        val sink: BufferedSink = Okio.buffer(Okio.sink(destination))
+        sink.writeAll(response.body().source())
+        sink.close()
+        return@fromCallable destination
+    }!!
```
Comment (nekdenis) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: avoid using !!

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+            .addAction(
+                    android.R.drawable.ic_menu_close_clear_cancel,
+                    context.getString(android.R.string.cancel),
+                    PendingIntent.getBroadcast(
+                            context,
+                            REQUEST_CANCEL,
+                            Intent(CANCEL_DOWNLOAD_ACTION),
+                            PendingIntent.FLAG_ONE_SHOT
+                    )
+            )!!
```
Comment (nekdenis) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: avoid using !!

File: app/src/main/java/com/stepango/archetype/player/loader/ProgressOkLoader.kt [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    }
+}
+
+class ProgressResponseBody(
+        val responseBody: ResponseBody,
+        val listener: ProgressUpdateListener) : ResponseBody() {
+
+    private var bufferedSource: BufferedSource? = null
+
+    override fun contentType() = responseBody.contentType()!!
```
Comment (nekdenis) [inline app/src/main/java/com/stepango/archetype/player/loader/ProgressOkLoader.kt]: avoid using !!

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+class EpisodeLoader(val service: Service) : ProgressUpdateListener {
+
+    val repo by lazyInject { episodesRepo() }
+    val toaster by lazyInject { toaster() }
+
+    val queueSubject = PublishSubject.create<Long>()!!
+    val notificationHelper = NotificationHelper(service, hashCode())
+    val loader = ProgressOkLoader(this)
+    val waitStack = ArrayList<Long>()
+    val dispatcher = queueSubject
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: dispatcher -> disposable?

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+            .timeout(10, TimeUnit.MINUTES)
+            .observeOn(Schedulers.io())
+            .subscribe(
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: use `bindSubscribe` with named parameters

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    fun stopForeground() {
+        service.stopForeground(true)
+    }
+
+    fun clearWaitStack() {
+        Observable
+                .fromIterable(waitStack)
+                .subscribe(
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: use `bindSubscribe` with named parameters
unsubscription?

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+                        { logger.e("error $it") }
+                )
+    }
+
+    fun queue(id: Long) {
+        Observable.just(queueSubject)
+                .doOnNext { updateEpisodeState(id, EpisodeDownloadState.WAIT) }
+                .doOnNext { waitStack.add(id) }
+                .observeOn(Schedulers.newThread())
+                .subscribe { it.onNext(id) }
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: use `bindSubscribe` with named parameters
unsubscription?

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    fun load(id: Long) {
+        repo.get(id)
+                .doOnSuccess { startForeground() }
+                .doOnSuccess { updateEpisodeState(it, EpisodeDownloadState.CANCEL) }
+                .doOnSuccess { waitStack.remove(id) }
+                .flatMap { startTask(it) }
+                .doAfterTerminate { stopForeground() }
+                .subscribe(
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: use `bindSubscribe` with named parameters
unsubscription?

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+                        { logger.d("done for episode id: $id") },
+                        { toaster.showToast("error on loading: $it") }
+                )
+    }
+
+    fun updateEpisodeState(episode: EpisodesModel, newState: EpisodeDownloadState) {
+        episode.state = newState
+        repo.save(episode.id, episode)
+                .subscribe()
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: unsubscription?

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    fun updateEpisodeState(id: Long, newState: EpisodeDownloadState) {
+        repo.get(id).subscribe( { updateEpisodeState(it, newState) }, { logger.e("error: $it") } )
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: unsubscription?
error handling?

File: app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    fun updateEpisodeFile(episode: EpisodesModel, file: File) {
+        episode.file = file.absolutePath
+        repo.save(episode.id, episode)
+                .subscribe()
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/loader/EpisodeLoaderService.kt]: unsubscription?
error handling?

File: app/src/main/java/com/stepango/archetype/player/data/wrappers/Wrappers.kt [hunk 19]
```diff
@@ -27,7 +27,7 @@ data class EpisodeWrapper(private val model: EpisodesModel) : AutoParcelable {
     val audioUrl: String = model.audioUrl
     val state: EpisodeDownloadState = model.state
     val file: String? = model.file
-    fun downloadButtonVisibility() = if (model.file == null) View.VISIBLE else View.GONE
+    fun downloadButtonVisibility(): Int = if (model.file == null) View.VISIBLE else View.INVISIBLE
```
Comment (stepango) [inline app/src/main/java/com/stepango/archetype/player/data/wrappers/Wrappers.kt]: replace with boolean, and use `visible` BindingAdapter

**Comment by stepango** [comment]
Resolve Codebeat issues https://codebeat.co/projects/github-com-stepango-archetype-master/pull_requests/47449

**Pull request merged**